### PR TITLE
[2163] Deprecate ProducerListener in favor of ProducerInterceptor

### PR DESCRIPTION
**Pull request by viktorardelean**
Resolves https://github.com/spring-projects/spring-kafka/issues/2163

Added Deprecated annotation where ProducerListener was used and suggested using ProducerInterceptor instead

Please let me know if there's anything to be changed.

Thanks

### CODE REVIEW

File: spring-kafka/src/main/java/org/springframework/kafka/support/CompositeProducerListener.java [hunk 1]
```diff
@@ -35,8 +35,10 @@
  * @author Artem Bilan
  *
  * @since 2.1.6
- *
+ * @deprecated since 3.0 in favor of {@link org.apache.kafka.clients.producer.ProducerInterceptor}
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/support/CompositeProducerListener.java:38]: Maybe "in favor of CompositeProducerInterceptor"  ?

File: spring-kafka/src/main/java/org/springframework/kafka/support/LoggingProducerListener.java [hunk 2]
```diff
@@ -32,7 +32,10 @@
  *
  * @author Marius Bogoevici
  * @author Gary Russell
+ * @deprecated since 3.0 in favor of {@link org.apache.kafka.clients.producer.ProducerInterceptor}
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/support/LoggingProducerListener.java:35]: Well, we don't have a `LoggingProducerInterceptor` to replace with.
I wonder if we need to implement one...

File: spring-kafka/src/main/java/org/springframework/kafka/support/ProducerListener.java [hunk 3]
```diff
@@ -36,7 +36,11 @@
  * @author Endika Gutiérrez
  *
  * @see org.apache.kafka.clients.producer.Callback
+ *
+ * @deprecated since 3.0 in favor of {@link org.apache.kafka.clients.producer.ProducerInterceptor}
  */
+
+@Deprecated(since = "3.0", forRemoval = true)
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/support/ProducerListener.java:43]: If we deprecate it, we have to replace its usage with the mentioned favor.
Any thoughts what could be done in the that `KafkaTemplate` for example?

I wonder if you ran your fix locally with the `./gradlew clean check`. It must show you that you still use some deprecate API.

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-kafka/src/main/java/org/springframework/kafka/support/CompositeProducerListener.java (hunk 1 above)
Comment (viktorardelean) [reply spring-kafka/src/main/java/org/springframework/kafka/support/CompositeProducerListener.java]: yes, I have modified

**Comment by garyrussell** [comment]
Thanks for the contribution; sorry for the delay.

After further review, I have decided not to deprecate `ProducerListener` because, unlike the `ProducerInterceptor`, it does not have access to the failed record in `onAcknowledgment()`.

**Comment by viktorardelean** [comment]
Thanks for the update @garyrussell !

**Pull request closed without merging**